### PR TITLE
Introducing Spiderlightning blog

### PR DESCRIPTION
**Pull request by Mossaka**

### CODE REVIEW

File: content/posts/introducing-slight.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+authorlink: "https://twitter.com/jiaxiao_zhou"
+image: "images/logos/twitter-card.png"
+tags: ["slight", "spiderlightning"]
+---
+
+At DeisLabs, we are researching and developing various WebAssembly tools and systems overtime. We have been working on a new project called SpiderLightning, which is a cloud system interface written in a Interface Definition Language (IDL) called [WIT](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md). We have also experimented with a CLI called slight, which we recently released "0.2.0" version. In this blog post, we will introduce SpiderLightning and slight.
+
+## Authoring portable distributed applications is hard
+
+Imagine you're a developer working at a consumer-facing company A. You built a web app that runs on on-premises data centers and used open-source projects to build data stores and pipelines. At beginning, there are only a few hundreds of customers, so the web app worked well. However, as the company grows, the number of customers increases rapidly. The web app becomes slow and unstable. Now you're tasked to migrate from on-premises data centers to public cloud.
```
Comment (danbugs) [inline content/posts/introducing-slight.md:16]:
```suggestion
Imagine you're a developer working at a consumer-facing company A. You built a web app that runs on on-premises data centers and used open-source projects to build data stores and pipelines. At the beginning, there are only a few hundreds of customers, so the web app worked well. However, as the company grows, the number of customers increases rapidly. The web app becomes slow and unstable. Now you're tasked to migrate from on-premises data centers to public cloud.
```

File: content/posts/introducing-slight.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+```bash
+slight -c slightfile.toml run -m target/wasm32-wasi/debug/spidey.wasm
+```
+
+If you see "Hello, SpiderLightning!", congratulations! You have successfully run your first slight application.
+
+## How does it work?
+
+You might noticed that slight takes two arguments. The first argument is a configuration file called `slightfile.toml`. The second argument is a WebAssembly module. They represent two phases of software development: operation and development, resp.
```
Comment (danbugs) [inline content/posts/introducing-slight.md:70]:
```suggestion
You might noticed that slight takes two arguments. The first argument is a configuration file called `slightfile.toml`. The second argument is a WebAssembly module. They represent two phases of software development: operation and development, respectively.
```

File: content/posts/introducing-slight.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+### Operations
+Let's take a look at the `slightfile.toml` file:
+
+```toml
+specversion = "0.2"
+
+[[capability]]
+resource = "kv.filesystem"
+name = "my-container"
+```
```
Comment (danbugs) [inline content/posts/introducing-slight.md:81]: I don't think this is the slightfile that you should get when running `slight new -n spidey@v0.1.0 rust` -- It should be `specversion = "0.1"`, iirc.

File: content/posts/introducing-slight.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+```toml
+specversion = "0.2"
+
+[[capability]]
+resource = "kv.redis"
+name = "my-container"
+    [capability.configs]
+    REDIS_ADDRESS = "redis://127.0.0.1:6379"
+```
```
Comment (danbugs) [inline content/posts/introducing-slight.md:95]: If you want to add capability configs like that, I'd recommend using `slight new -n spidey@v0.2.0 rust`

**Review by danbugs — CHANGES_REQUESTED** [review]
~mostly LGTM, great initiative! There are just a couple of changes I think are needed here and there.

File: content/posts/introducing-slight.md (hunk 1 above)
Comment (devigned) [reply content/posts/introducing-slight.md]: I think we need to work on this paragraph a bit as it's the hook that will get folks to continue reading.

I'm not clear what building data stores and pipelines are. I mean, I can kind of imagine what is going on, but it isn't super clear to me. Consider something really simple the application does. Maybe it handling airline reservations or some other use case that is easily relatable and less tech focused.

Also consider cutting anything extra from the paragraph. One sentence that jumps out at me is "The web app becomes slow and unstable." I don't believe that slight / SL will magically make a poorly written / scaling application into a well written, linearly scalable application. Try to focus on the problem that you are solving, making the application more portable and have lower cost of running due to reduced resource requirements.

Comment (devigned) [reply content/posts/introducing-slight.md]: The next paragraph is the great for describing the agnostic APIs. Lean more into that.

File: content/posts/introducing-slight.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+Migrating to the cloud means that you will need to re-write the code to use cloud vendor resources, such as datastores and pipelines. Even worse, the company strategically decided to use two cloud vendors to reduce the risk of vendor lock-in. Now you need to re-write the code twice. This is a common scenario for many companies. The problem is that the code is not portable.
+
+## What is SpiderLightning and slight?
+
+Making code more portable has always been a challenge. Think about POSIX. It is a system interface designed to maintain compatibility between operating systems. The key is to design a system interface that abstracts away cloud provider-specific capabilities.
+
+SpiderLightning is a cloud system interface that is designed to make applications portable across cloud providers.
```
Comment (devigned) [inline content/posts/introducing-slight.md:24]:
```suggestion
SpiderLightning is a distributed application system interface that is designed to make applications portable across hosting providers (cloud, on-premise, IoT, etc).
```

File: content/posts/introducing-slight.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+It's interface defines a set of common capabilities that are often needed to build distributed applications. These capabilities include key-value stores, message queue, pub/sub, runtime configuration etc.
```
Comment (devigned) [inline content/posts/introducing-slight.md:26]:
```suggestion
Its interface defines a set of common capabilities that are often needed to build distributed applications. These capabilities include key-value stores, message queue, pub/sub, runtime configuration etc.
```

File: content/posts/introducing-slight.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+	// delete the payload for a given key.
+	delete: func(key:string) -> expected<unit, error>
+}
+```
+
+Only at runtime, the WebAssembly runtime will provide the implementation of the key-value store interface. At runtime, slight knows which key-value store is used by looking at the configuration file.
+
+## Standardization
+
+Without standardization and community support, SpiderLightning will not go far. We are working with the WebAssembly community to standardize these distributed application capability interfaces.
```
Comment (devigned) [inline content/posts/introducing-slight.md:145]: Try not to focus on SpiderLightning's success. The success that we want to see more than anything is the success of the community around the interfaces. Success via standardization will provide a common set of capabilities accessible by ANY runtime that exposes those interfaces.

File: content/posts/introducing-slight.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+In fact, they are already in WebAssembly System Interface (WASI) phase 1 proposals. You can find the proposals [here](https://github.com/WebAssembly/WASI/blob/main/Proposals.md).
+
+Our next step is to work with multiple parties to design a set of interfaces that follow the least common denominator (LCD) principle.
```
Comment (devigned) [inline content/posts/introducing-slight.md:149]:
```suggestion
Our next step is to work with multiple parties to design a set of interfaces that find the balance between rich feature sets and portability across providers. 
```

File: content/posts/introducing-slight.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+
+Without standardization and community support, SpiderLightning will not go far. We are working with the WebAssembly community to standardize these distributed application capability interfaces.
+
+In fact, they are already in WebAssembly System Interface (WASI) phase 1 proposals. You can find the proposals [here](https://github.com/WebAssembly/WASI/blob/main/Proposals.md).
+
+Our next step is to work with multiple parties to design a set of interfaces that follow the least common denominator (LCD) principle. 
+
+## Finishing thoughts
+
+We are excited for work like WebAssembly Component Model and WASI to mature. We believe that in the future, you will be able to build portable and secure wasm applications that can run on any places, including cloud, edge, and IoT.
```
Comment (devigned) [inline content/posts/introducing-slight.md:153]:
```suggestion
We are excited for the WebAssembly Component Model and WASI to continue maturing. We believe that in the future, you will be able to build portable and secure wasm applications that can run nearly anywhere, including on-premise, cloud, edge, IoT, on satellites in space, anywhere!
```

File: content/posts/introducing-slight.md [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+description: "Announcing SpiderLightning and slight, a distributed application runtime for Wasm featuring portable application building blocks."
+date: 2022-10-31 00:00:00 +0000 UTC
+authorname: "Jiaxiao (Joe) Zhou"
+author: "@jiaxiao_zhou"
+authorlink: "https://twitter.com/jiaxiao_zhou"
+image: "images/logos/twitter-card.png"
+tags: ["slight", "spiderlightning", "Wasm", "WASI"]
+---
+
+At DeisLabs, we are researching and developing various WebAssembly tools and systems overtime. We have been working on a new project called SpiderLightning, which is a distributed application interface written in an Interface Definition Language (IDL) called [WIT](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md). SpiderLightning is paired with a CLI called slight, which we recently released "0.2.0" version. In this blog post, we will introduce SpiderLightning and slight.
```
Comment (squillace) [inline content/posts/introducing-slight.md:12]: `...various WebAssembly tools and infrastructure.`

the overtime part doesn't add anything.

Comment (squillace) [inline content/posts/introducing-slight.md:12]: `....which is a set of interfaces and implementations intended to provide developers in the WebAssembly ecosystem a way to use the WebAssembly component model quickly and easily for distributed applications that need only a core set of service functionality. The interfaces are written in an Interface Definition Language (IDL) called [WIT](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md), part of the WebAssembly component model work in the [Bytecode Alliance Foundation (BCA)](https://bytecodealliance.org). The SpiderLightning project also comes with a CLI called slight, and we recently released "0.2.0" version. In this blog post, we will introduce SpiderLightning and slight.`

File: content/posts/introducing-slight.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+## Authoring portable distributed applications is hard
+
+Imagine you're a developer working at a consumer-facing company. You built a web app that runs on on-premises data centers and used open-source projects to build data stores and pipelines. At the beginning, there are only a few hundreds of customers, so the web app worked well. However, as the company grows, the number of customers increases rapidly. The web app becomes slow and unstable. This is a classic scaling issue. Now you're tasked to migrate from on-premises data centers to public cloud because they will horizontal scale the application for you.
```
Comment (squillace) [inline content/posts/introducing-slight.md:16]: the first thing that will happen with this framing is that the reader will say.... uh, scaling has nothing to do with this. you just buy more computers. Why, they will ask, does the web app become slow and unstable? 

So there's a reader disconnect right out of the gate. There are MILLIONS of projects that make "web apps run faster" but you don't even mention anything ABOUT spiderlight that means "the problem of scaling is generally solved", right? 

So FRAME it a bit better in the introduction: what does WEBASSEMBLY do that containers can't do. THEN you can say hey, if you were building in containers, you would not be able to take advantage of wasm features. So the following conversation is about how if you want to take advantage of the WebAssembly component model to ease, and therefore scale out your ability to make wasm apps fast and well for the edge AND hyperscale..... something like this is required up front.

keep all this you have, of course. you'll get to it! But without the introduction that's better, stronger, and has more positioning for the reader much of the reaction will be disbelief that you couldn't have just bought more machines -- problem solved!!!

**Review by squillace — CHANGES_REQUESTED** [review]
I left the critical thing that must be addressed otherwise the remainder of the post is going to be dismissed out of hand by well more than half your readers.

File: content/posts/introducing-slight.md (hunk 11 above)
Comment (Mossaka) [reply content/posts/introducing-slight.md]: resolved offline

**Review by squillace — APPROVED** [review]

File: content/posts/introducing-spiderlightning-and-slight.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/deislabs/spiderlightning/main/install.sh)"
+```
+
+After installing slight, you can create your very first slight application by running the following command:
+
+```bash
+slight new -n spidey@v0.2.0 rust && cd spidey
+```
+
+This command will create a new rust project called spidey. The project will use the SpiderLightning interface version 0.2.0. We also support C.
```
Comment (danbugs) [inline content/posts/introducing-spiderlightning-and-slight.md:53]:
```suggestion
This command will create a new Rust project called spidey that will use the SpiderLightning interfaces at version 0.2.0. Other than Rust, we also support starting C projects.
```

**Review by danbugs — APPROVED** [review]
LGTM~ just one small suggestion!

**Pull request merged**